File: PlusOne.py
class Solution:
    # @param digits, a list of integer digits
    # @return a list of integer digits
    def plusOne(self, digits):
        digits = digits[::-1]
        digits[0] += 1
        carry = 0
        for i in xrange(len(digits)):
            digits[i] += carry
            if digits[i] >= 10:
                digits[i] = digits[i] % 10
                carry = 1
            else:
                carry = 0
                break
        if carry:
            digits.append(1)
        return digits[::-1]

    # not reverse, 6 times faster
    def plusOne2(self, digits):
        carry = 0
        for i in xrange(len(digits)-1, -1, -1):
            digits[i] += carry
            if digits[i] >= 10:
                digits[i] = digits[i] % 10
                carry = 1
            else:
                carry = 0
                break
        if carry:
            digits.insert(0, 1)
        return digits
                    
            

File: CPP/Second/plus_one.cc
// https://oj.leetcode.com/problems/plus-one/

// Given a non-negative number represented as an array of digits, plus one to the
// The digits are stored such that the most significant digit is at the head of

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

/*
    Time O(n)
*/
class Solution {
public:
    vector<int> plusOne(vector<int> &digits) {
        int carry = 1;
        for (int i = digits.size() - 1; i >= 0; i--) {
            if (carry == 0)
                break;
            digits[i] += carry;
            carry = digits[i] / 10;
            digits[i] %= 10;
        }
        if (carry)
            digits.insert(digits.begin(), 1);
        return digits;
    }
};
int main(int argc, char *argv[])
{

    return 0;
}
